middleware/pagination_test.go

```diff
@@ -0,0 +1,291 @@
+			})
+
+			router.ServeHTTP(rr, req)
+		},
+		Entry("Null data",
+			10,
+			0,
+			nil,
+			0,
+			"",
+			"",
+			"",
+			"",
+			nil,
+			fmt.Errorf("invalid data set: data cannot be nil"),
+		),
+		Entry("Empty data",
+			10,
+			0,
+			[]string{},
+			0,
+			"/test?limit=10&offset=0",
+			"/test?limit=10&offset=0",
+			"",
+			"",
+			[]interface{}{},
+			nil,
+		),
+		Entry("Data with 1 item",
+			10,
+			0,
+			[]string{"test"},
+			1,
+			"/test?limit=10&offset=0",
+			"/test?limit=10&offset=0",
+			"",
+			"",
+			[]string{"test"},
+			nil,
+		),
+		Entry("Data with 10 items",
+			10,
+			0,
+			[]string{"test", "test", "test", "test", "test", "test", "test", "test", "test", "test"},
+			10,
+			"/test?limit=10&offset=0",
+			"/test?limit=10&offset=0",
+			"",
+			"",
+			[]string{"test", "test", "test", "test", "test", "test", "test", "test", "test", "test"},
+			nil,
+		),
+		Entry("Data with 11 items",
+			10,
+			0,
+			[]string{"test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test"},
+			11,
+			"/test?limit=10&offset=0",
+			"/test?limit=10&offset=1",
+			"/test?limit=10&offset=10",
+			"",
+			[]string{"test", "test", "test", "test", "test", "test", "test", "test", "test", "test"},
+			nil,
+		),
+		Entry("Data with 12 items",
+			10,
+			0,
+			[]string{"test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test"},
+			12,
+			"/test?limit=10&offset=0",
+			"/test?limit=10&offset=2",
+			"/test?limit=10&offset=10",
+			"",
+			[]string{"test", "test", "test", "test", "test", "test", "test", "test", "test", "test"},
+			nil,
+		),
+		Entry("Data with 20 items",
+			10,
+			0,
+			[]string{"test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test", "test"},
+			20,
+			"/test?limit=10&offset=0",
+			"/test?limit=10&offset=10",
+			"/test?limit=10&offset=10",
+			"",
+			[]string{"test", "test", "test", "test", "test", "test", "test", "test", "test", "test"},
+			nil,
+		),
+		Entry("Data with 20 items and offset 10",
+			10,
+			10,
+			[]string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen", "twenty"},
+			20,
+			"/test?limit=10&offset=0",
+			"/test?limit=10&offset=10",
+			"",
+			"/test?limit=10&offset=0",
+			[]string{"eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen", "twenty"},
+			nil,
+		),
+		Entry("Data with offset greater than data length",
+			10,
+			100,
+			[]string{"one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen", "twenty"},
+			20,
+			"/test?limit=10&offset=0",
+			"/test?limit=10&offset=10",
+			"",
+			"/test?limit=10&offset=90",
```

Comment:
In this case, I wonder if we should return the previous link as an empty string/nil?